File: lab3/sol/mpi_derived_types_sol.f90
!
!  STL file format 
!
!  UINT8[80] – Header
!  UINT32 – Number of triangles
!
!  foreach triangle
!  REAL32[3] – Normal vector
!  REAL32[3] – Vertex 1
!  REAL32[3] – Vertex 2
!  REAL32[3] – Vertex 3
!  UINT16 – Attribute byte count
!  end
!
!  (see https://en.wikipedia.org/wiki/STL_(file_format)
!
module types
  integer, parameter :: dp = kind(0.0d0)
  integer, parameter :: sp = kind(0.0)
end module types

module stl
  use mpi
  use types
  implicit none

  type :: stl_triangle_t
     real(kind=sp) :: n(3)
     real(kind=sp) :: v1(3)
     real(kind=sp) :: v2(3)
     real(kind=sp) :: v3(3)
     integer(kind=2) :: attrib
  end type stl_triangle_t

  type :: stl_model_t
     character(len=80) :: hdr
     integer :: n_tri
     type(stl_triangle_t), allocatable :: tri(:)
  end type stl_model_t

  integer :: MPI_STL_TRI
  
contains

  subroutine stl_read(fname, model)
    character(len=*), intent(in) :: fname
    type(stl_model_t), intent(inout) :: model
    integer(kind=MPI_OFFSET_KIND) :: byte_offset
    integer :: fh, ierr, pe_size, pe_rank, stl_offset, n_tri
    integer :: csize, isize, tsize

    call MPI_Comm_size(MPI_COMM_WORLD, pe_size, ierr)
    call MPI_Comm_rank(MPI_COMM_WORLD, pe_rank, ierr)

    call MPI_Type_size(MPI_CHARACTER, csize, ierr)
    call MPI_Type_size(MPI_INTEGER, isize, ierr)
    call MPI_Type_size(MPI_STL_TRI, tsize, ierr)

    call MPI_File_open(MPI_COMM_WORLD, fname, &
         MPI_MODE_RDONLY, MPI_INFO_NULL, fh, ierr)
    
    if (pe_rank .eq. 0) write(*,*) 'Reading STL file: ', fname
    ! Read STL header
    call MPI_File_read_all(fh, model%hdr, 80, MPI_CHARACTER, &
         MPI_STATUS_IGNORE, ierr)
    
    ! Make sure it's a binary STL file 
    if (model%hdr(1:6) .eq. 'solid') then
       write(*,*) 'ASCII STL files not supported!'
       stop 
    end if
    
    ! Read how many triangles the file contains 
    call MPI_File_read_all(fh, n_tri, 1, MPI_INTEGER, &
         MPI_STATUS_IGNORE, ierr)
    if (pe_rank .eq. 0) write(*,*) 'Found: ', n_tri,' triangles'

    ! Compute how many triangles this rank should read
    model%n_tri = (n_tri + pe_size - pe_rank - 1) / pe_size

    ! Allocate memory for triangles, and read them
    allocate(model%tri(model%n_tri))    

    ! Compute the offset into the list of triangles
    stl_offset = 0
    call MPI_Exscan(model%n_tri, stl_offset, 1, MPI_INTEGER, &
         MPI_SUM, MPI_COMM_WORLD, ierr)

    ! Compute offset into the file
    byte_offset = 80 * csize + isize + stl_offset * tsize

    call MPI_File_read_at_all(fh, byte_offset, model%tri, model%n_tri, MPI_STL_TRI, &
         MPI_STATUS_IGNORE, ierr)

    call MPI_File_close(fh, ierr)
    if (pe_rank .eq. 0) write(*,*) 'Done'

  end subroutine stl_read

  subroutine stl_write(fname, model)
    character(len=*), intent(in) :: fname
    type(stl_model_t), intent(in) :: model
    integer(kind=MPI_OFFSET_KIND) :: byte_offset
    integer :: fh, ierr, pe_size, pe_rank, stl_offset, n_tri
    integer :: csize, isize, tsize
    
    call MPI_Comm_size(MPI_COMM_WORLD, pe_size, ierr)
    call MPI_Comm_rank(MPI_COMM_WORLD, pe_rank, ierr)

    call MPI_Type_size(MPI_CHARACTER, csize, ierr)
    call MPI_Type_size(MPI_INTEGER, isize, ierr)
    call MPI_Type_size(MPI_STL_TRI, tsize, ierr)
    
    call MPI_File_open(MPI_COMM_WORLD, fname, &
         MPI_MODE_WRONLY + MPI_MODE_CREATE, MPI_INFO_NULL, fh, ierr)
    
    if (pe_rank .eq. 0) write(*,*) 'Writing STL file: ', fname

    ! Write STL header
    call MPI_File_write_all(fh, model%hdr, 80, MPI_CHARACTER, &
         MPI_STATUS_IGNORE, ierr)

    ! Compute the total number of triangles 
    call MPI_Allreduce(model%n_tri, n_tri, 1, MPI_INTEGER, &
         MPI_SUM, MPI_COMM_WORLD, ierr)

    ! Write number of triangles
    call MPI_File_write_all(fh, n_tri, 1, MPI_INTEGER, &
         MPI_STATUS_IGNORE, ierr)

    ! Compute the offset into the list of triangles
    stl_offset = 0
    call MPI_Exscan(model%n_tri, stl_offset, 1, MPI_INTEGER, &
         MPI_SUM, MPI_COMM_WORLD, ierr)

    ! Compute the offset into the file
    byte_offset = 80 * csize + isize + stl_offset * tsize
    
    ! Write all triangles
    call MPI_File_write_at_all(fh, byte_offset, model%tri, model%n_tri, MPI_STL_TRI, &
         MPI_STATUS_IGNORE, ierr)

    call MPI_File_close(fh, ierr)
    if (pe_rank .eq. 0) write(*,*) 'Done'

  end subroutine stl_write
end module stl

program mpi_derived_types
  use stl
  use mpi
  implicit none
  type(stl_model_t) :: model
  type(stl_triangle_t) :: triangle
  integer(kind=MPI_ADDRESS_KIND) :: disp(5), base
  integer :: type(5), len(5), ierr

  call MPI_Init(ierr)

  ! Setup displacement of each block in the struct
  call MPI_Get_address(triangle%n, disp(1), ierr)
  call MPI_Get_address(triangle%v1, disp(2), ierr)
  call MPI_Get_address(triangle%v2, disp(3), ierr)
  call MPI_Get_address(triangle%v3, disp(4), ierr)
  call MPI_Get_address(triangle%attrib, disp(5), ierr)
  
  base = disp(1) 
  disp(1) = disp(1) - base 
  disp(2) = disp(2) - base 
  disp(3) = disp(3) - base 
  disp(4) = disp(4) - base 
  disp(5) = disp(5) - base 
  
  len = 3
  len(5) = 1
  
  type(1:4) = MPI_REAL
  type(5) = MPI_INTEGER2
  
  ! Create dervied MPI type
  call MPI_Type_create_struct(5, len, disp, type, MPI_STL_TRI, ierr)
  call MPI_Type_commit(MPI_STL_TRI, ierr)

  call stl_read("./data/sphere.stl", model)
  call stl_write("out.stl", model)
  deallocate(model%tri)

  call MPI_Finalize(ierr)

end program mpi_derived_types

File: lab3/sol/mpi_derived_types_sol.c
/*
  STL file format 

  UINT8[80] – Header
  UINT32 – Number of triangles

  foreach triangle
  REAL32[3] – Normal vector
  REAL32[3] – Vertex 1
  REAL32[3] – Vertex 2
  REAL32[3] – Vertex 3
  UINT16 – Attribute byte count
  end

  (see https://en.wikipedia.org/wiki/STL_(file_format)
  
*/

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <mpi.h>

#define STL_HDR_SIZE 80
MPI_Datatype MPI_STL_TRI;

typedef struct {
  float n[3];			/* Normal vector */
  float v1[3];			/* Vertex 1 */
  float v2[3];			/* Vertex 2 */
  float v3[3];			/* Vertex 3 */
  uint16_t attrib;		/* Attribute byte count */
} __attribute__((packed)) stl_triangle_t;

typedef struct {
  char hdr[STL_HDR_SIZE];	/* Header */
  uint32_t n_tri;		/* Number of triangles */
  stl_triangle_t *tri;		/* Triangles */
} stl_model_t;


void stl_read(const char *fname, stl_model_t *model) {
  MPI_File fh;
  int pe_size, pe_rank;
  uint n_tri, stl_offset;
  MPI_Offset byte_offset;

  MPI_Comm_size(MPI_COMM_WORLD, &pe_size);
  MPI_Comm_rank(MPI_COMM_WORLD, &pe_rank);

  MPI_File_open(MPI_COMM_WORLD, fname, MPI_MODE_RDONLY, MPI_INFO_NULL, &fh);
  if (pe_rank == 0) printf("Reading STL file: %s\n", fname);

  /* Read STL header */
  MPI_File_read_all(fh, model->hdr, STL_HDR_SIZE, MPI_CHAR, MPI_STATUS_IGNORE);

  /* Make sure it's a binary STL file */
  if (strncmp(model->hdr, "solid", 5) == 0) {
    fprintf(stderr, "ASCII STL files not supported!\n");
    exit(-1);
  }

  /* Read how many triangles the file contains */
  MPI_File_read_all(fh, &n_tri, 1, MPI_UNSIGNED, MPI_STATUS_IGNORE);
  if (pe_rank == 0) printf("Found: %d triangles\n", n_tri);

  /* Compute how many triangles this rank should read */
  model->n_tri = (n_tri + pe_size - pe_rank - 1) / pe_size;

  /* Allocate memory for triangles, and read them */
  model->tri = malloc(model->n_tri * sizeof(stl_triangle_t));

  /* Compute the offset into the list of triangles */
  stl_offset = 0;
  MPI_Exscan(&model->n_tri, &stl_offset, 1, 
	     MPI_UNSIGNED, MPI_SUM, MPI_COMM_WORLD);

  /* Compute offset into the file */
  byte_offset = STL_HDR_SIZE * sizeof(char) + sizeof(uint32_t) + 
    stl_offset * sizeof(stl_triangle_t);

  MPI_File_read_at_all(fh, byte_offset, model->tri, model->n_tri,
		       MPI_STL_TRI, MPI_STATUS_IGNORE);

  MPI_File_close(&fh);
  if (pe_rank == 0) printf("Done\n");

}

void stl_write(const char *fname, stl_model_t *model) {
  MPI_File fh;
  int pe_size, pe_rank;
  uint n_tri, stl_offset;
  MPI_Offset byte_offset;

  MPI_Comm_size(MPI_COMM_WORLD, &pe_size);
  MPI_Comm_rank(MPI_COMM_WORLD, &pe_rank);
  
  MPI_File_open(MPI_COMM_WORLD, fname,
		MPI_MODE_WRONLY | MPI_MODE_CREATE, MPI_INFO_NULL, &fh);
  if (pe_rank == 0) printf("Writing STL file: %s\n", fname);

  /* Write STL header */
  MPI_File_write_all(fh, model->hdr, STL_HDR_SIZE, MPI_CHAR, MPI_STATUS_IGNORE);

  /* Compute the total number of triangles */
  MPI_Allreduce(&model->n_tri, &n_tri, 1,
		MPI_UNSIGNED, MPI_SUM, MPI_COMM_WORLD);

  /* Write number of triangles */
  MPI_File_write_all(fh, &n_tri, 1, MPI_UNSIGNED, MPI_STATUS_IGNORE);


  /* Compute the offset into the list of triangles */
  stl_offset = 0;
  MPI_Exscan(&model->n_tri, &stl_offset, 1, 
	     MPI_UNSIGNED, MPI_SUM, MPI_COMM_WORLD);

  /* Compute offset into the file */
  byte_offset = STL_HDR_SIZE * sizeof(char) + sizeof(uint32_t) + 
    stl_offset * sizeof(stl_triangle_t);
  
  /* Write all triangles */
  MPI_File_write_at_all(fh, byte_offset, model->tri, model->n_tri, 
  			MPI_STL_TRI, MPI_STATUS_IGNORE);

  MPI_File_close(&fh);
  if (pe_rank == 0) printf("Done\n");
}

int main(int argc, char **argv) {
  int i;
  stl_model_t model;
  stl_triangle_t triangle;

  const int len[5] = {3, 3, 3, 3, 1};
  const MPI_Datatype type[5] = {MPI_FLOAT, MPI_FLOAT, 
				MPI_FLOAT, MPI_FLOAT, MPI_UNSIGNED_SHORT};
  MPI_Datatype MPI_STL_TRI_STRUCT;
  MPI_Aint base, disp[5], sizeofstruct;

  MPI_Init(&argc, &argv);

  /* Setup displacement of each block in the struct */
  MPI_Get_address(&triangle.n[0], disp); 
  MPI_Get_address(&triangle.v1[0], disp + 1); 
  MPI_Get_address(&triangle.v2[0], disp + 2); 
  MPI_Get_address(&triangle.v3[0], disp + 3); 
  MPI_Get_address(&triangle.attrib, disp + 4); 

  base = disp[0]; 
  for (i=0; i < 5; i++) disp[i] = MPI_Aint_diff(disp[i], base); 
  MPI_Type_create_struct(5, len, disp, type, &MPI_STL_TRI_STRUCT);

  /* Resize the dervied type to account for padding by the compiler */
  MPI_Get_address(&triangle+1, &sizeofstruct); 
  sizeofstruct = MPI_Aint_diff(sizeofstruct, base); 
  MPI_Type_create_resized(MPI_STL_TRI_STRUCT, 0, sizeofstruct, &MPI_STL_TRI);
 
  /* Create dervied MPI type */
  MPI_Type_commit(&MPI_STL_TRI);


  stl_read("./data/sphere.stl", &model);
  stl_write("out.stl", &model);
  free(model.tri);

  MPI_Finalize();
  
  return 0;
}


